### PR TITLE
Highlight query results on judgment page

### PR DESCRIPTION
**Pull request by timcowlishaw**
## Changes in this PR:

This adds highlighting of query terms in the judgment body, and a UI for navigating between them, in order that users can better identify at a glance how relevant a judgment is to their query, and also navigate to relevant sections easily.

We also switch to using `waffle` for feature flags instead of `django-flags`, and put this behind a feature flag - **the flag `highlight_search_query` will need to be created and enabled on staging and production to make this visible**.

Tests will pass once [v17.2.0 of the API client is released](https://github.com/nationalarchives/ds-caselaw-custom-api-client/pull/448) and the SCSS will build once [this shared frontend change](https://github.com/nationalarchives/ds-caselaw-frontend/pull/17) is merged!

## Trello card / Rollbar error (etc)

https://trello.com/c/tnpX6eJW/1412-search-highlight-term-within-a-judgment

#### Before
Screenshot 2023-11-09 at 08 14 05
#### After
Screenshot 2023-11-09 at 08 13 17

#### Before
Screenshot 2023-11-09 at 08 14 13

#### After
Screenshot 2023-11-09 at 08 13 34

### CODE REVIEW

**Comment by timcowlishaw** [comment]
[SEARCH: Highlight term within a judgment](https://trello.com/c/tnpX6eJW/1412-search-highlight-term-within-a-judgment)

**Comment by timcowlishaw** [comment]
Couple of good codeclimate suggestions there - will do a quick refactor tomorrow AM

**Comment by timcowlishaw** [comment]
> Couple of good codeclimate suggestions there - will do a quick refactor tomorrow AM

Actually, i take that back - I think the explicitness of the current version would be lost if i refactored, and it's not a huge amount of repetition. I vote to leave as is, unless anyone objects?

File: pyproject.toml [hunk 1]
```diff
@@ -3,3 +3,4 @@
 ignore="H021,H023,H030,H031"
 profile="django"
 indent=2
+custom_blocks="flag"
```
Comment (jacksonj04) [inline pyproject.toml:6]: \o/

File: judgments/migrations/0002_alter_courtdates_options.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Generated by Django 3.2.22 on 2023-11-09 05:34
```
Comment (jacksonj04) [inline judgments/migrations/0002_alter_courtdates_options.py:1]: This is a bit concerning; where does this change come from?

**Review by jacksonj04 — CHANGES_REQUESTED** [review]

File: judgments/views/detail.py [hunk 3]
```diff
@@ -75,16 +80,17 @@ def get_best_pdf(request, document_uri):
 
 
 def detail(request, document_uri):
+    query = request.GET.get("query")
     document = get_published_document_by_uri(document_uri)
 
     # If the document_uri which was requested isn't the canonical URI of the document, redirect the user
     if document_uri != document.uri:
```
[the remark this answers is not in the document]
Comment (timcowlishaw) [reply judgments/views/detail.py:87]: aha good shout - death to unneeded complexity! I'll remove.

File: judgments/migrations/0002_alter_courtdates_options.py (hunk 2 above)
Comment (timcowlishaw) [reply judgments/migrations/0002_alter_courtdates_options.py:1]: ok, i think there was an earlier change somewhere on `main` which would have generated a migration, but for whatever reason we never ran the migration generation command at the time, so therefore when we added waffle and ran the migrations for that, it generated and applied the old migration too. Maybe this should come out of this PR and I'll open a new one for it on its own?

File: ds_judgements_public_ui/static/js/src/modules/document_navigation_links.js [hunk 4]
```diff
@@ -45,18 +105,109 @@ $(() => {
...
+                root: null,
+                rootMargin: `-25% 0px -25% 0px`,
+            },
+        );
+
+        $("#start-of-document-link").text("Start");
+        $("#end-of-document-link").text("End");
+    }
+
+    $("#end-of-document-link").on("click", (event) => {
```
[the remark this answers is not in the document]
Comment (timcowlishaw) [reply ds_judgements_public_ui/static/js/src/modules/document_navigation_links.js:185]: oh yep, that's a lovely solution - nothing had occured to me that didn't make the JS a little too obscure while also removing the dupication. I'll do that this AM!

File: config/settings/base.py [hunk 5]
```diff
@@ -72,15 +69,18 @@
...
+    "waffle",
 ]
 
 LOCAL_APPS = [
     # Your stuff: custom apps go here
     "judgments.apps.JudgmentsConfig",
 ]
 # https://docs.djangoproject.com/en/dev/ref/settings/#installed-apps
-INSTALLED_APPS = DJANGO_APPS + LOCAL_APPS
+INSTALLED_APPS = DJANGO_APPS + THIRD_PARTY_APPS + LOCAL_APPS
```
[the remark this answers is not in the document]
Comment (timcowlishaw) [reply config/settings/base.py:83]: 🙌

File: judgments/migrations/0002_alter_courtdates_options.py (hunk 2 above)
Comment (timcowlishaw) [reply judgments/migrations/0002_alter_courtdates_options.py]: As discussed, i'll open another PR with that change!

**Review by jacksonj04 — APPROVED** [review]
This gets the green light to merge and deploy, _but_ we shouldn't set up the feature flag on PUI in any kind of deployment mode (ie only opt in specific users via admin) until we've had the changes made to the cookie banner.

**Pull request merged**